beamer/agent/chain.py

```diff
@@ -625,59 +624,65 @@ def maybe_prove(claim: Claim, context: Context) -> None:
         context.target_chain.rpc_url,
         context.source_chain.rpc_url,
         context.config.account.key,
-        proof_tx,
+        proved_tx,
         True,
     )
 
     def on_future_done(f: Future) -> None:
         try:
             f.result()
         except Exception as ex:
-            context.logger.error("Optimism prove failed", ex=ex, tx_hash=proof_tx)
+            context.logger.error("Optimism prove failed", ex=ex, tx_hash=proved_tx)
         else:
-            claim.message_proved = True
+            assert request is not None
+            claim.proved_tx = proved_tx
+            if request.fill_tx == proved_tx:
+                request.fill_timestamp = Timestamp(int(time.time()))
+            elif claim.invalidation_tx == proved_tx:
+                claim.invalidation_timestamp = Timestamp(int(time.time()))
         finally:
-            context.l1_resolutions.pop(proof_tx, None)
+            assert proved_tx is not None
+            context.l1_resolutions.pop(proved_tx, None)
 
     future.add_done_callback(on_future_done)
-    context.l1_resolutions[proof_tx] = future
+    context.l1_resolutions[proved_tx] = future
 
-    context.logger.info("Initiated Optimism prove", request=request, claim=claim, tx_hash=proof_tx)
+    context.logger.info(
+        "Initiated Optimism prove", request=request, claim=claim, tx_hash=proved_tx
+    )
 
 
 def maybe_resolve(claim: Claim, context: Context) -> bool:
     request = context.requests.get(claim.request_id)
 
     assert request is not None, "Active claim for non-existent request"
 
-    if not claim.message_proved:
+    if claim.proved_tx is None:
         return False
 
     if not _proof_ready_for_l1_relay(request) and not _invalidation_ready_for_l1_relay(claim):
         return False
 
     timestamp_finalized = False
 
-    if request.fill_timestamp:
+    if request.fill_tx == claim.proved_tx:
+        assert request.fill_timestamp is not None
         timestamp_finalized = _timestamp_is_l1_finalized(
             request.fill_timestamp, context, request.target_chain_id
         )
-        assert request.fill_tx is not None
-        proof_tx = request.fill_tx
-    elif claim.invalidation_timestamp:
+    else:
+        assert claim.invalidation_timestamp is not None
         timestamp_finalized = _timestamp_is_l1_finalized(
             claim.invalidation_timestamp, context, request.target_chain_id
         )
-        assert claim.invalidation_tx is not None
-        proof_tx = claim.invalidation_tx
 
     if not timestamp_finalized:
         return False
 
     if not _l1_resolution_threshold_reached(claim, context):
         return False
 
-    if proof_tx in context.l1_resolutions:
+    if claim.proved_tx in context.l1_resolutions:
         return False
```

Comment:
I would move this check earlier, e.g. right after we check that `claim.proved_tx` is not `None`. We could even combine it with that check.